x/claim/keeper/claim.go

```diff
@@ -0,0 +1,235 @@
+	return nil
+}
+
+// SetClaimables set claimable amount from balances object
+func (k Keeper) SetClaimRecords(ctx sdk.Context, claimRecords []types.ClaimRecord) error {
+	for _, claimRecord := range claimRecords {
+		err := k.SetClaimRecord(ctx, claimRecord)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// GetClaimables get claimables for genesis export
+func (k Keeper) GetClaimRecords(ctx sdk.Context, chain types.Chain) []types.ClaimRecord {
+	store := ctx.KVStore(k.storeKey)
+	prefixStore := prefix.NewStore(store, chainToStorePrefix(chain))
+
+	iterator := prefixStore.Iterator(nil, nil)
+	defer iterator.Close()
+
+	claimRecords := []types.ClaimRecord{}
+	for ; iterator.Valid(); iterator.Next() {
+
+		claimRecord := types.ClaimRecord{}
+
+		err := proto.Unmarshal(iterator.Value(), &claimRecord)
+		if err != nil {
+			panic(err)
+		}
+
+		claimRecords = append(claimRecords, claimRecord)
+	}
+	return claimRecords
+}
+
+// GetClaimRecord returns the claim record for a specific address
+func (k Keeper) GetClaimRecord(ctx sdk.Context, addr string, chain types.Chain) (types.ClaimRecord, error) {
+	store := ctx.KVStore(k.storeKey)
+	prefixStore := prefix.NewStore(store, chainToStorePrefix(chain))
+	addrBytes := []byte(strings.ToLower(addr))
+	if !prefixStore.Has(addrBytes) {
+		return types.ClaimRecord{}, nil
+	}
+	bz := prefixStore.Get(addrBytes)
+
+	claimRecord := types.ClaimRecord{}
+	err := proto.Unmarshal(bz, &claimRecord)
+	if err != nil {
+		return types.ClaimRecord{}, err
+	}
+
+	return claimRecord, nil
+}
+
+// GetUserTotalClaimable returns the total claimable amount for an address
+func (k Keeper) GetUserTotalClaimable(ctx sdk.Context, addr string, chain types.Chain) (sdk.Coins, error) {
+	claimRecord, err := k.GetClaimRecord(ctx, addr, chain)
+	if err != nil {
+		return sdk.Coins{}, err
+	}
+	if claimRecord.Address == "" {
+		return sdk.Coins{}, nil
+	}
+
+	totalClaimable := sdk.Coins{}
+	for action := range types.Action_name {
+		claimableForAction, err := k.GetClaimableAmountForAction(ctx, addr, types.Action(action), chain)
+		if err != nil {
+			return sdk.Coins{}, err
+		}
+		totalClaimable = totalClaimable.Add(claimableForAction...)
+	}
+	return totalClaimable, nil
+}
+
+// GetClaimable returns claimable amount for a specific action done by an address
+func (k Keeper) GetClaimableAmountForAction(ctx sdk.Context, addr string, action types.Action, chain types.Chain) (sdk.Coins, error) {
+	claimRecord, err := k.GetClaimRecord(ctx, addr, chain)
+	if err != nil {
+		return nil, err
+	}
+
+	if claimRecord.Address == "" {
+		return sdk.Coins{}, nil
+	}
+
+	// if action already completed, nothing is claimable
+	if claimRecord.ActionCompleted[action] {
+		return sdk.Coins{}, nil
+	}
+
+	params := k.GetParams(ctx)
+
+	// If we are before the start time, do nothing.
+	// This case _shouldn't_ occur on chain, since the
+	// start time ought to be chain start time.
+	if ctx.BlockTime().Before(params.AirdropStartTime) {
```

Comment:
usually I would use block height , instead of a specfic `time.Time`

Reply:
Understood, I think this is more of a product question and currently they have specified a time, we can certainly revisit, but I don't think its a critical difference.